pkg/collector/corechecks/servicediscovery/module/impl_linux.go

```diff
@@ -613,6 +616,87 @@ func (s *discovery) updateServicesCPUStats(services []model.Service) error {
 	return nil
 }
 
+func (s *discovery) enrichContainerData(service *model.Service, containers map[string]*agentPayload.Container, pidToCid map[int]string) {
+	id, ok := pidToCid[service.PID]
+	if !ok {
+		return
+	}
+
+	// The tags we look for service name generation, in their priority order.
+	// The map entries will be filled as we go through the containers tags.
+	tagsPriority := []struct {
+		tagName  string
+		tagValue *string
+	}{
+		{"service", nil},
+		{"app", nil},
+		{"short_image", nil},
+		{"kube_container_name", nil},
+		{"kube_deployment", nil},
+		{"kube_service", nil},
+	}
+
+	service.ContainerID = id
+	log.Debugf("Found container id for %v: %v", service.Name, id)
+
+	// We got the service name from container tags before, no need to do it again.
+	if service.GeneratedNameFromContainerTags {
+		return
+	}
+
+	container, ok := containers[id]
+	if !ok {
+		return
+	}
+
+	var serviceName string
+
+	for _, tag := range container.Tags {
+		// Get index of separator between name and value
+		sepIndex := strings.IndexRune(tag, ':')
+		if sepIndex < 0 || sepIndex >= len(tag)-1 {
+			// Malformed tag; we skip it
+			continue
+		}
+
+		for i := range tagsPriority {
+			if tag[:sepIndex] != tagsPriority[i].tagName {
+				// Not a tag we care about; we skip it
+				continue
+			}
+
+			value := tag[sepIndex+1:]
+			tagsPriority[i].tagValue = &value
+			break
+		}
+	}
+
+	for _, tag := range tagsPriority {
+		if tag.tagValue == nil {
+			continue
+		}
+
+		serviceName = *tag.tagValue
+		log.Debugf("Using %v:%v tag for service name", tag.tagName, *tag.tagValue)
+		break
+	}
+
+	if serviceName != "" {
+		s.mux.Lock()
+		defer s.mux.Unlock()
+
+		serviceInfo, ok := s.cache[int32(service.PID)]
+		if !ok {
+			return
+		}
+
+		service.GeneratedName = serviceName
```

Comment:
Setting `service` doesn't need the lock, and in fact should be done regardless of the cache check since the `service` should probably be enhanced even if the `serviceInfo` is gone from the cache (due to parallel deletion).